source/common/upstream/load_balancer_impl.cc

```diff
@@ -615,41 +631,150 @@ bool LoadBalancerBase::isHostSetInPanic(const HostSet& host_set) const {
   return false;
 }
 
-void ZoneAwareLoadBalancerBase::calculateLocalityPercentage(
-    const HostsPerLocality& hosts_per_locality, uint64_t* ret) {
-  uint64_t total_hosts = 0;
-  for (const auto& locality_hosts : hosts_per_locality.get()) {
-    total_hosts += locality_hosts.size();
+ZoneAwareLoadBalancerBase::LocalityPercentagesPtr
+ZoneAwareLoadBalancerBase::calculateLocalityPercentages(
+    const HostsPerLocality& local_hosts_per_locality,
+    const HostsPerLocality& upstream_hosts_per_locality) {
```

Comment:
This method is now more complicated to determine how to interleave the localities between the local and upstream host sets even with mismatches and differing numbers of zones.

We use the `locality` field of each individual host in the hosts per locality to determine a valid interleaving, assuming that localities are already sorted properly (as described in `HostsPerLocalityImpl`).

The strategy is akin to the "merge" step in mergesort where we iteratively process entries out of each host set until all localities have been merged.